### PR TITLE
ci: Use Docker for CI on Linux

### PR DESCRIPTION
**Pull request by jkayani**
### Description

Dockerize the CI environment on Linux to avoid this problem: https://github.com/IBM/argocd-vault-plugin/pull/193#issuecomment-911876575. 

This solution works based on this Actions run https://github.com/IBM/argocd-vault-plugin/runs/3508366010?check_suite_focus=true which makes these changes atop PR 193 to prove we avoid the special-case of Azure auth solely via cloud-provider metadata endpoints

Specific CI changes:

- Split the CI matrix since we use Docker on Linux not on Mac
- Add a Dockerfile that downloads the Golang deps at build time
- Run `go vet` only on our code (nor our deps) and run `go test` in a network isolated container to avoid successful Azure auth
- Upload code coverage only once
- Remove some un-used stuff from the Makefile

### Checklist
Please make sure that your PR fulfills the following requirements:
- [ ] Reviewed the guidelines for contributing to this repository
- [ ] The commit message follows the [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] Tests for the changes have been updated
- [ ] Docs have been added / updated

### Type of Change

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [x] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

### CODE REVIEW

File: .github/workflows/pipeline.yml [hunk 1]
```diff
@@ -1,20 +1,38 @@
 name: Pipeline
 
 on:
-  push:
-    branches: [main]
```
Comment (werne2j) [inline .github/workflows/pipeline.yml:5]: Why not run this on a push to main?

File: .github/workflows/pipeline.yml [hunk 2]
```diff
@@ -1,20 +1,38 @@
...
   pull_request:
     # The branches below must be a subset of the branches above
     branches: [main, next]
 
 jobs:
-  CI:
-    name: CI
+  linux:
```
Comment (werne2j) [inline .github/workflows/pipeline.yml:9]: Interesting to split them up by OS. Could we try adding windows back?

File: .github/workflows/pipeline.yml (hunk 1 above)
Comment (jkayani) [reply .github/workflows/pipeline.yml:5]: We can but I didn't see the point - we know the tests will pass and how the coverage will look from the results of the PR that the push to main came from, right? I can put it back just didn't really see the point

File: .github/workflows/pipeline.yml (hunk 2 above)
Comment (jkayani) [reply .github/workflows/pipeline.yml:9]: Sure, though I don't see the value of that unless we start publishing Windows binaries, which we could certainly do.  But IIRC we had problems with the tests on Windows a long time ago. Also the Azure-specific behavior applies to workloads in Windows and Linux VMs so we'd have to run in Docker on Windows which would make things even slower

Comment (werne2j) [reply .github/workflows/pipeline.yml:9]: Yeah, you're right, we are not doing windows binaries so no need

**Review by werne2j — APPROVED** [review]

**Pull request merged**